File: answers/lintcode/Binary Tree Level Order Traversal.cpp
//@level Medium 
//@type Queue Binary Tree Binary Tree Traversal Breadth First Search
//@result Accepted Total Runtime: 17 ms 100% test cases passed.

/**
 * Definition of TreeNode:
 * class TreeNode {
 * public:
 *     int val;
 *     TreeNode *left, *right;
 *     TreeNode(int val) {
 *         this->val = val;
 *         this->left = this->right = NULL;
 *     }
 * }
 */
 
 
class Solution {
    /**
     * @param root: The root of binary tree.
     * @return: Level order a list of lists of integer
     */
public:
    vector<vector<int> > levelOrder(TreeNode *root) {
        // write your code here
		answer.clear();
		levelOrder(root, 0);
		return answer;
    }
    void levelOrder(TreeNode *root, int level) {
		if (NULL == root) {
			return;
		}
		if (level >= answer.size()) {
			answer.resize(level + 1, vector<int> ());
		}
		answer[level].push_back(root->val);
		levelOrder(root->left, level + 1);
		levelOrder(root->right, level + 1);
	}
	vector<vector<int> > answer;
};

File: answers/leetcode/Binary Tree Level Order Traversal/Binary Tree Level Order Traversal.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
class Solution {
    public List<List<Integer>> levelOrder(TreeNode root) {
        List<List<Integer>> result = new ArrayList<List<Integer>>();
        this.dfs(root, 0, result);
        return result;
    }

    private void dfs(TreeNode root, int level, List<List<Integer>> result) {
        if (null == root) {
            return;
        }
        while (result.size() <= level) {
            result.add(new ArrayList<>());
        }
        result.get(level).add(root.val);
        this.dfs(root.left, level + 1, result);
        this.dfs(root.right, level + 1, result);
    }
}